### PR TITLE
Tomandersen/multi db

### PR DESCRIPTION
**Pull request by tom-andersen**
Add Firestore named DB support.

### CODE REVIEW

File: .changeset/heavy-spiders-fry.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+'@firebase/firestore': minor
+'@firebase/firestore-compat': patch
```
Comment (wu-hui) [inline .changeset/heavy-spiders-fry.md:3]: Does this really affect compat?

Comment (wu-hui) [reply .changeset/heavy-spiders-fry.md:3]: Hmm, OK it's for the console build, got it.

File: packages/firestore/src/api/database.ts [hunk 2]
```diff
@@ -173,17 +174,57 @@ export function initializeFirestore(
...
+/**
+ * Returns the existing {@link Firestore} instance that is associated with the
+ * default {@link @firebase/app#FirebaseApp}. If no instance exists, initializes a new
+ * instance with default settings.
+ *
+ * @returns The {@link Firestore} instance of the provided app.
+ */
+export function getFirestore(): Firestore;
+/**
+ * Returns the existing {@link Firestore} instance that is associated with the
```
Comment (wu-hui) [inline packages/firestore/src/api/database.ts:186]: `the existing default Firestore instance`?

File: packages/firestore/src/api/database.ts [hunk 3]
```diff
@@ -173,17 +174,57 @@ export function initializeFirestore(
   return provider.initialize({ options: settings });
 }
 
+/**
+ * Returns the existing {@link Firestore} instance that is associated with the
```
Comment (wu-hui) [inline packages/firestore/src/api/database.ts:178]: `the existing default Firestore instance`? Same for lite api.

File: packages/firestore/test/lite/integration.test.ts [hunk 4]
```diff
@@ -106,14 +106,56 @@ describe('Firestore', () => {
...
+
+  it('returns different instance from named app', () => {
+    const app = initializeApp(
+      { apiKey: 'fake-api-key', projectId: 'test-project' },
+      'test-app-getFirestore'
+    );
+    const fs1 = getFirestore(app);
+    const fs2 = getFirestore(app, 'name1');
+    const fs3 = getFirestore(app, 'name2');
+    expect(fs1).to.not.be.equal(fs2).and.equal(fs3);
```
Comment (wu-hui) [inline packages/firestore/test/lite/integration.test.ts:128]: This line is confusing, should fs1 equal to fs3 or not? Please split.

File: packages/firestore/test/lite/integration.test.ts [hunk 5]
```diff
@@ -106,14 +106,56 @@ describe('Firestore', () => {
...
+      'test-app-getFirestore-1'
+    );
+    const app2 = initializeApp(
+      { apiKey: 'fake-api-key', projectId: 'test-project' },
+      'test-app-getFirestore-2'
+    );
+    const fs1 = getFirestore();
+    const fs2 = getFirestore(app1);
+    const fs3 = getFirestore(app2);
+    expect(fs1).to.not.be.equal(fs2).and.equal(fs3);
```
Comment (wu-hui) [inline packages/firestore/test/lite/integration.test.ts:157]: Same as above.

File: packages/firestore/test/lite/integration.test.ts [hunk 6]
```diff
@@ -106,14 +106,56 @@ describe('Firestore', () => {
     expect(fs1).to.be.an.instanceOf(Firestore);
   });
 
-  it('returns same instance', () => {
+  it('returns same default instance from named app', () => {
```
Comment (wu-hui) [inline packages/firestore/test/lite/integration.test.ts:109]: Please also add some tests to `test/integration/api/database.test.ts`. No need to be this thorough, just sanity checks.

Comment (wu-hui) [inline packages/firestore/test/lite/integration.test.ts:109]: Request to add:

* Test some real actions (queries) for the created dbs.
* Create a db with 'name1' and defaultApp, write some document and terminate it. Create another db with 'name1' and explicitly specifies defaultApp, make sure you can read the previous document from cache. Ideally also a db with 'name2' and make sure data is not leaking.

**Review by wu-hui — COMMENTED** [review]
Nice PR. Mostly nits and requests to add more tests.

File: packages/firestore/test/lite/integration.test.ts (hunk 6 above)
Comment (tom-andersen) [reply packages/firestore/test/lite/integration.test.ts:109]: Add tests. IN doing so, some problem were revealed. Fixed.

Comment (tom-andersen) [reply packages/firestore/test/lite/integration.test.ts:109]: Done.

File: packages/firestore/test/integration/util/helpers.ts [hunk 7]
```diff
@@ -182,7 +184,40 @@ export async function withTestDbsSettings(
...
+    }
+  }
+}
+
+export async function withNamedTestDbs(
+  persistence: boolean,
+  dbNames: string[],
+  fn: (db: Firestore[]) => Promise<void>
+): Promise<void> {
+  // Named DBs requires preparing project with DBs beforehand.
```
Comment (wu-hui) [inline packages/firestore/test/integration/util/helpers.ts:211]: This comment does not read well..please fix.

**Review by wu-hui — APPROVED** [review]

File: packages/firestore/test/integration/util/helpers.ts (hunk 7 above)
Comment (tom-andersen) [reply packages/firestore/test/integration/util/helpers.ts]: Fixed comment

**Pull request merged**